File: src/Billing/BroadcastService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\{Database,Outbox};
final class BroadcastService
{
    public function __construct(private Database $db, private Outbox $outbox) {}
    /** Create a broadcast job. Returns the row. */
    public function create(string $targetType, string $text, string $adminId, string $adminName, string $category = 'system'): array
    {
        if (mb_strlen($text) < 1 || mb_strlen($text) > 4000) throw new BillingError('Текст рассылки: 1–4000 символов.');
        if (!in_array($targetType, ['all','active','inactive','paid','trial','telegram','email'], true)) throw new BillingError('Некорректный сегмент.');
        $id = Database::id();
        $now = time();
        $this->db->execute(
            'INSERT INTO broadcast_history(id,target_type,message_text,total_count,status,admin_id,admin_name,category,created_at) VALUES(?,?,?,?,?,?,?,?,?)',
            [$id, $targetType, $text, 0, 'in_progress', $adminId, $adminName, $category, $now]
        );
        $this->outbox->enqueue('broadcast.run', 'broadcast:'.$id.':1', ['broadcast_id' => $id]);
        $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $adminId, 'broadcast.created', $id, $now]);
        return $this->db->one('SELECT * FROM broadcast_history WHERE id=?', [$id]);
    }
    /** How many per-user broadcast sends to enqueue per broadcast.run tick. */
    private static int $chunk = 150;
    /** Delay (s) before the next broadcast.run is processed, so a chunk drains before the next is scheduled. */
    private static int $chunkDelay = 45;

    /** Run a broadcast: pick recipients and enqueue per-user sends in throttled chunks. */
    public function run(string $broadcastId): void
    {
        $b = $this->db->one('SELECT * FROM broadcast_history WHERE id=?', [$broadcastId]);
        if (!$b || !in_array($b['status'], ['in_progress', 'running'], true)) return;
        if ($b['status'] === 'in_progress') {
            $recipients = $this->recipients($b['target_type']);
            $this->db->execute('UPDATE broadcast_history SET total_count=? WHERE id=?', [count($recipients), $broadcastId]);
            $this->db->execute("UPDATE broadcast_history SET status='running' WHERE id=?", [$broadcastId]);
        } else {
            $recipients = $this->recipients($b['target_type']);
        }
        $remaining = array_values(array_filter(
            $recipients,
            fn($tgId) => !$this->db->one('SELECT 1 FROM outbox WHERE dedup_key=?', ['bsend:'.$broadcastId.':'.$tgId])
        ));
        $enqueued = 0;
        foreach ($remaining as $tgId) {
            if ($enqueued >= self::$chunk) break;
            $this->outbox->enqueue('broadcast.send', 'bsend:'.$broadcastId.':'.$tgId, ['broadcast_id' => $broadcastId, 'chat_id' => $tgId, 'text' => $b['message_text']]);
            $enqueued++;
        }
        if ($enqueued === 0) return;
        // More recipients remain: schedule the next chunk after a pause instead of flooding
        // the queue. Dedup keeps every recipient queued exactly once across ticks; each
        // continuation tick gets its own dedup key (broadcast:<id>:<n>) because the first
        // tick key is consumed (done) once the worker processes it.
        if (count($remaining) > $enqueued) {
            $this->outbox->enqueue('broadcast.run', 'broadcast:'.$broadcastId.':'.(string)$this->nextBatchKey($broadcastId), ['broadcast_id' => $broadcastId], self::$chunkDelay);
        }
    }
    private function nextBatchKey(string $broadcastId): int
    {
        $row = $this->db->one("SELECT count(*) c FROM outbox WHERE topic='broadcast.run' AND dedup_key LIKE ?", ['broadcast:'.$broadcastId.':%']);
        return (int)($row['c'] ?? 0) + 1;
    }
    private function recipients(string $target): array
    {
        $now = time();
        $rows = match ($target) {
            'all' => $this->db->all("SELECT telegram_id FROM users WHERE telegram_id IS NOT NULL AND disabled=0"),
            'active' => $this->db->all("SELECT DISTINCT u.telegram_id FROM users u JOIN subscriptions s ON s.user_id=u.id WHERE u.telegram_id IS NOT NULL AND u.disabled=0 AND s.status IN ('active','trial') AND s.expires_at>?", [$now]),
            'inactive' => $this->db->all("SELECT telegram_id FROM users WHERE telegram_id IS NOT NULL AND disabled=0 AND id NOT IN (SELECT user_id FROM subscriptions WHERE status IN ('active','trial') AND expires_at>?)", [$now]),
            'paid' => $this->db->all("SELECT telegram_id FROM users WHERE telegram_id IS NOT NULL AND disabled=0 AND has_had_paid_subscription=1"),
            'trial' => $this->db->all("SELECT DISTINCT u.telegram_id FROM users u JOIN subscriptions s ON s.user_id=u.id WHERE u.telegram_id IS NOT NULL AND u.disabled=0 AND s.is_trial=1 AND s.status IN ('active','trial')"),
            'telegram' => $this->db->all("SELECT telegram_id FROM users WHERE telegram_id IS NOT NULL AND disabled=0"),
            default => [],
        };
        return array_values(array_filter(array_map(fn($r) => (string)($r['telegram_id'] ?? ''), $rows), fn($v) => $v !== ''));
    }
    /** Mark a single send as delivered or failed. */
    public function markSent(string $broadcastId, bool $ok): void
    {
        $this->db->execute('UPDATE broadcast_history SET sent_count=sent_count+1 WHERE id=?', [$broadcastId]);
        if (!$ok) $this->db->execute('UPDATE broadcast_history SET failed_count=failed_count+1 WHERE id=?', [$broadcastId]);
        $b = $this->db->one('SELECT * FROM broadcast_history WHERE id=?', [$broadcastId]);
        if ($b && (int)$b['sent_count'] + (int)$b['failed_count'] >= (int)$b['total_count']) {
            $this->db->execute("UPDATE broadcast_history SET status='completed',completed_at=? WHERE id=?", [time(), $broadcastId]);
        }
    }
    public function list(int $limit = 50): array
    {
        return $this->db->all('SELECT * FROM broadcast_history ORDER BY created_at DESC LIMIT ?', [$limit]);
    }
}

File: rails_backend/app/services/billing/broadcast_service.rb
module Billing
  class BroadcastService
    CHUNK_SIZE = 150
    CHUNK_DELAY = 45
    TARGETS = %w[all active inactive paid trial telegram email].freeze

    def initialize(outbox: Infrastructure::OutboxService.new)
      @outbox = outbox
    end

    def create(target_type:, text:, admin_id:, admin_name:, category: "system")
      text = text.to_s
      raise BillingError, "Текст рассылки: 1–4000 символов." unless text.length.between?(1, 4000)
      raise BillingError, "Некорректный сегмент." unless TARGETS.include?(target_type)
      now = Time.now.to_i
      ApplicationRecord.transaction do
        row = BroadcastHistory.create!(id: Infrastructure::IdGenerator.call, target_type: target_type,
          message_text: text, total_count: 0, status: "in_progress", admin_id: admin_id,
          admin_name: admin_name, category: category, created_at: now)
        @outbox.enqueue("broadcast.run", "broadcast:#{row.id}:1", { broadcast_id: row.id })
        audit(admin_id, "broadcast.created", row.id, now)
        row
      end
    end

    def run(broadcast_id)
      broadcast = BroadcastHistory.find_by(id: broadcast_id)
      return unless broadcast && %w[in_progress running].include?(broadcast.status)
      recipients = recipients_for(broadcast.target_type)
      if broadcast.status == "in_progress"
        broadcast.update!(total_count: recipients.length, status: "running")
      end
      remaining = recipients.reject do |chat_id|
        OutboxJob.exists?(dedup_key: "bsend:#{broadcast_id}:#{chat_id}") ||
          BroadcastDelivery.exists?(broadcast_id: broadcast_id, chat_id: chat_id)
      end
      batch = remaining.first(CHUNK_SIZE)
      batch.each do |chat_id|
        @outbox.enqueue("broadcast.send", "bsend:#{broadcast_id}:#{chat_id}",
          { broadcast_id: broadcast_id, chat_id: chat_id, text: broadcast.message_text })
      end
      if batch.empty?
        complete_if_finished(broadcast_id)
      elsif remaining.length > batch.length
        next_batch = OutboxJob.where(topic: "broadcast.run").where("dedup_key LIKE ?", "broadcast:#{broadcast_id}:%").count + 1
        @outbox.enqueue("broadcast.run", "broadcast:#{broadcast_id}:#{next_batch}",
          { broadcast_id: broadcast_id }, delay: CHUNK_DELAY)
      end
    end

    def mark_sent(broadcast_id, chat_id, ok)
      ApplicationRecord.transaction do
        now = Time.now.to_i
        inserted = BroadcastDelivery.insert_all(
          [{ broadcast_id: broadcast_id, chat_id: chat_id, outcome: ok ? "sent" : "failed", completed_at: now }],
          unique_by: %i[broadcast_id chat_id]
        ).count
        next if inserted.zero?
        column = ok ? :sent_count : :failed_count
        BroadcastHistory.where(id: broadcast_id).update_all("#{column} = #{column} + 1")
        complete_if_finished(broadcast_id)
      end
    end

    def list(limit: 50)
      BroadcastHistory.order(created_at: :desc).limit([[limit.to_i, 1].max, 100].min)
    end

    private

    def recipients_for(target)
      now = Time.now.to_i
      users = User.where(disabled: 0).where.not(telegram_id: [nil, ""])
      case target
      when "active"
        users.joins(:subscriptions).where(subscriptions: { status: %w[active trial] })
          .where("subscriptions.expires_at > ?", now).distinct.pluck(:telegram_id)
      when "inactive"
        users.where.not(id: Subscription.where(status: %w[active trial]).where("expires_at > ?", now).select(:user_id)).pluck(:telegram_id)
      when "paid"
        users.where(has_had_paid_subscription: 1).pluck(:telegram_id)
      when "trial"
        users.joins(:subscriptions).where(subscriptions: { is_trial: 1, status: %w[active trial] })
          .where("subscriptions.expires_at > ?", now).distinct.pluck(:telegram_id)
      when "all", "telegram"
        users.pluck(:telegram_id)
      else
        []
      end.compact.map(&:to_s).reject(&:blank?).uniq
    end

    def complete_if_finished(id)
      row = BroadcastHistory.find_by(id: id)
      return unless row && row.status == "running" && row.sent_count.to_i + row.failed_count.to_i >= row.total_count.to_i
      row.update!(status: "completed", completed_at: Time.now.to_i)
    end

    def audit(actor, action, subject, now)
      AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor, action: action, subject: subject, created_at: now)
    end
  end
end
